enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,307 @@ package hit_tracker_client
+}
+
+func (h *NoOpHitTracker) TrackMiss(d *repb.Digest) error {
+	return nil
 }
 
-func (h HitTrackerFactory) NewACHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
-	return &HitTracker{}
+func (h *NoOpHitTracker) TrackDownload(d *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-func (h HitTrackerFactory) NewCASHitTracker(ctx context.Context, requestMetadata *repb.RequestMetadata) interfaces.HitTracker {
-	return &HitTracker{}
+func (h *NoOpHitTracker) TrackUpload(d *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
 }
 
-func (h *HitTracker) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+type NoOpTransferTimer struct {
 }
 
-func (h *HitTracker) TrackMiss(d *repb.Digest) error {
+func (t *NoOpTransferTimer) CloseWithBytesTransferred(bytesTransferredCache, bytesTransferredClient int64, compressor repb.Compressor_Value, serverLabel string) error {
 	return nil
 }
 
+func (t *NoOpTransferTimer) Record(bytesTransferred int64, duration time.Duration, compressor repb.Compressor_Value) error {
+	return nil
+}
+
+type HitTrackerClient struct {
+	ctx             context.Context
+	enqueueFn       func(context.Context, *repb.RequestMetadata, *hitpb.CacheHit)
+	client          hitpb.HitTrackerServiceClient
+	requestMetadata *repb.RequestMetadata
+}
+
+// This should only be used for Action Cache hit tracking.
+func (h *HitTrackerClient) SetExecutedActionMetadata(md *repb.ExecutedActionMetadata) {
+	alert.UnexpectedEvent("Unexpected call to SetExecutedActionMetadata")
+}
+
+// Local cache misses hit the backing cache, which will take care of
+// hit-tracking for this request.
+func (h *HitTrackerClient) TrackMiss(d *repb.Digest) error {
+	return nil
+}
+
+func (h *HitTrackerFactory) groupID(ctx context.Context) groupID {
+	claims, err := claims.ClaimsFromContext(ctx)
+	if err != nil {
+		return interfaces.AuthAnonymousUser
+	}
+	return groupID(claims.GetGroupID())
+}
+
+func (h *HitTrackerFactory) enqueue(ctx context.Context, requestMetadata *repb.RequestMetadata, hit *hitpb.CacheHit) {
+	requestMetadata = proto.Clone(requestMetadata).(*repb.RequestMetadata)
+
+	if h.stopped.Load() {
+		log.Warning("hit_tracker_client.enqueue after worker shutdown, sending RPC synchronously")
+		hit.RequestMetadata = requestMetadata
+		if _, err := h.client.Track(ctx, &hitpb.TrackRequest{Hits: []*hitpb.CacheHit{hit}}); err != nil {
+			log.Infof("Error sending HitTrackerService.Track RPC: %v", err)
+		}
+		return
+	}
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	groupID := h.groupID(ctx)
+	if _, ok := h.hitsByGroup[groupID]; !ok {
+		h.hitsByGroup[groupID] = &cacheHits{
+			hits: make([]*hitpb.CacheHit, 0, initialSliceCapacity),
+		}
+	}
+
+	groupHits := h.hitsByGroup[groupID]
+	if len(groupHits.hits) >= h.maxPendingHitsPerGroup {
+		log.CtxWarning(ctx, "Exceeded maximum number of pending cache hits, dropping some.")
```

Comment:
UnexpectedEvent so we get alerted here? I don't think it's likely we'll hit the 100K limit here but if we do hit it, we want to address it quickly.

Reply:
Done